API Test Failures

Multiple failures in the main API tests:

```
Running tool: /usr/local/bin/go test -tags test -timeout 30s github.com/jamiefdhurst/journal

2018/07/01 10:45:32 Serving on http://127.0.0.1:60131
2018/07/01 10:45:32 GET: /api/v1/post
2018/07/01 10:45:32 GET: /api/v1/post/test
2018/07/01 10:45:32 GET: /api/v1/post/random
2018/07/01 10:45:32 POST: /api/v1/post
--- FAIL: TestApiV1Create (0.01s)
	/Users/jamie/Sites/_go/src/github.com/jamiefdhurst/journal/journal_test.go:138: Expected 200 status code
	/Users/jamie/Sites/_go/src/github.com/jamiefdhurst/journal/journal_test.go:147: Expected:
			{"id":4,"slug":"test-4","title":"Test 4","date":"2018-06-01T00:00:00Z","content":"<p>Test 4!</p>"}
		Got:
			
		<!DOCTYPE html>
		<html lang="en">
		<head>
		    <meta charset="UTF-8" />
		    <title>Jamie's Journal</title>
		    <meta name="viewport" content="device-width" />
		
		    <link rel="stylesheet" type="text/css" href="/css/default.min.css" />
		</head>
		<body>
		    <div id="container">
		        <header role="banner">
		            <h1>Jamie's Journal</h1>
		        </header>
		
		        
		
		<h2>Page Not Found</h2>
		
		<p><a href="/">Go Home</a></p>
		
		
		    </div>
		    <script src="/js/default.min.js"></script>
		</body>
		</html>
2018/07/01 10:45:32 POST: /api/v1/post
--- FAIL: TestApiV1Create_InvalidRequest (0.01s)
	/Users/jamie/Sites/_go/src/github.com/jamiefdhurst/journal/journal_test.go:163: Expected 400 status code
2018/07/01 10:45:32 POST: /api/v1/post
--- FAIL: TestApiV1Create_MissingData (0.01s)
	/Users/jamie/Sites/_go/src/github.com/jamiefdhurst/journal/journal_test.go:179: Expected 400 status code
2018/07/01 10:45:32 PUT: /api/v1/post/test
--- FAIL: TestApiV1Update (0.01s)
	/Users/jamie/Sites/_go/src/github.com/jamiefdhurst/journal/journal_test.go:195: Expected 200 status code
	/Users/jamie/Sites/_go/src/github.com/jamiefdhurst/journal/journal_test.go:204: Expected:
			{"id":1,"slug":"test","title":"A different title","date":"2018-01-01T00:00:00Z","content":"<p>Test!</p>"}
		Got:
			
		<!DOCTYPE html>
		<html lang="en">
		<head>
		    <meta charset="UTF-8" />
		    <title>Jamie's Journal</title>
		    <meta name="viewport" content="device-width" />
		
		    <link rel="stylesheet" type="text/css" href="/css/default.min.css" />
		</head>
		<body>
		    <div id="container">
		        <header role="banner">
		            <h1>Jamie's Journal</h1>
		        </header>
		
		        
		
		<h2>Page Not Found</h2>
		
		<p><a href="/">Go Home</a></p>
		
		
		    </div>
		    <script src="/js/default.min.js"></script>
		</body>
		</html>
2018/07/01 10:45:32 PUT: /api/v1/post/random
2018/07/01 10:45:32 PUT: /api/v1/post/test
--- FAIL: TestApiV1Update_InvalidRequest (0.01s)
	/Users/jamie/Sites/_go/src/github.com/jamiefdhurst/journal/journal_test.go:236: Expected 400 status code
FAIL
FAIL	github.com/jamiefdhurst/journal	0.065s
Error: Tests failed.
```

--- a/internal/app/controller/apiv1/create_test.go
+++ b/internal/app/controller/apiv1/create_test.go
@@ -13,6 +13,8 @@ import (
 
 func TestCreate_Run(t *testing.T) {
 	db := &database.MockSqlite{}
+	db.Result = &database.MockResult{}
+	db.Rows = &database.MockRowsEmpty{}
 	container := &app.Container{Db: db}
 	response := controller.NewMockResponse()
 	response.Reset()

--- a/internal/app/controller/web/index.go
+++ b/internal/app/controller/web/index.go
@@ -22,6 +22,7 @@ func (c *Index) Run(response http.ResponseWriter, request *http.Request) {
 	js := model.Journals{Container: c.Super.Container.(*app.Container), Gs: &model.Giphys{Container: c.Super.Container.(*app.Container)}}
 	c.Journals = js.FetchAll()
 	query := request.URL.Query()
+	c.Saved = false
 	if query["saved"] != nil {
 		c.Saved = true
 	}

--- a/internal/app/controller/web/index_test.go
+++ b/internal/app/controller/web/index_test.go
@@ -34,4 +34,10 @@ func TestIndex_Run(t *testing.T) {
 	if !strings.Contains(response.Content, "Journal saved") {
 		t.Error("Expected saved banner to be displayed on screen")
 	}
+	response.Reset()
+	request, _ = http.NewRequest("GET", "/", strings.NewReader(""))
+	controller.Run(response, request)
+	if strings.Contains(response.Content, "Journal saved") {
+		t.Error("Expected saved banner to be hidden, but it is showing")
+	}
 }

--- a/internal/app/controller/web/new.go
+++ b/internal/app/controller/web/new.go
@@ -36,7 +36,7 @@ func (c *New) Run(response http.ResponseWriter, request *http.Request) {
 		}
 
 		js := model.Journals{Container: c.Super.Container.(*app.Container), Gs: &model.Giphys{Container: c.Super.Container.(*app.Container)}}
-		journal := model.Journal{ID: 0, Slug: model.Slugify(request.FormValue("title")), Title: model.Slugify(request.FormValue("title")), Date: request.FormValue("date"), Content: request.FormValue("content")}
+		journal := model.Journal{ID: 0, Slug: model.Slugify(request.FormValue("title")), Title: request.FormValue("title"), Date: request.FormValue("date"), Content: request.FormValue("content")}
 		js.Save(journal)
 
 		http.Redirect(response, request, "/?saved=1", 302)

--- a/internal/app/controller/web/new_test.go
+++ b/internal/app/controller/web/new_test.go
@@ -13,6 +13,8 @@ import (
 
 func TestNew_Run(t *testing.T) {
 	db := &database.MockSqlite{}
+	db.Result = &database.MockResult{}
+	db.Rows = &database.MockRowsEmpty{}
 	container := &app.Container{Db: db}
 	response := controller.NewMockResponse()
 	controller := &New{}

--- a/internal/app/model/journal.go
+++ b/internal/app/model/journal.go
@@ -62,6 +62,21 @@ func (js *Journals) CreateTable() error {
 	return err
 }
 
+// EnsureUniqueSlug Make sure the current slug is unique
+func (js *Journals) EnsureUniqueSlug(slug string, addition int) string {
+	newSlug := slug
+	if addition > 0 {
+		newSlug = strings.Join([]string{slug, "-", strconv.Itoa(addition)}, "")
+	}
+	exists := js.FindBySlug(newSlug)
+	if exists.ID > 0 {
+		addition++
+		return js.EnsureUniqueSlug(slug, addition)
+	}
+
+	return newSlug
+}
+
 // FetchAll Get all journals
 func (js *Journals) FetchAll() []Journal {
 	rows, err := js.Container.Db.Query("SELECT * FROM `" + journalTable + "` ORDER BY `date` DESC")
@@ -94,8 +109,12 @@ func (js *Journals) Save(j Journal) Journal {
 
 	// Convert content for saving
 	j.Content = js.Gs.ExtractContentsAndSearchAPI(j.Content)
+	if j.Slug == "" {
+		j.Slug = Slugify(j.Title)
+	}
 
 	if j.ID == 0 {
+		j.Slug = js.EnsureUniqueSlug(j.Slug, 0)
 		res, _ = js.Container.Db.Exec("INSERT INTO `"+journalTable+"` (`slug`, `title`, `date`, `content`) VALUES(?,?,?,?)", j.Slug, j.Title, j.Date, j.Content)
 	} else {
 		res, _ = js.Container.Db.Exec("UPDATE `"+journalTable+"` SET `slug` = ?, `title` = ?, `date` = ?, `content` = ? WHERE `id` = ?", j.Slug, j.Title, j.Date, j.Content, strconv.Itoa(j.ID))

--- a/internal/app/model/journal_test.go
+++ b/internal/app/model/journal_test.go
@@ -58,6 +58,35 @@ func TestJournals_CreateTable(t *testing.T) {
 	}
 }
 
+func TestJournals_EnsureUniqueSlug(t *testing.T) {
+	db := &database.MockSqlite{}
+	db.ErrorMode = false
+	container := &app.Container{Db: db}
+	js := Journals{Container: container}
+
+	// Test no result
+	db.Rows = &database.MockRowsEmpty{}
+	actual := js.EnsureUniqueSlug("test", 0)
+	if actual != "test" {
+		t.Errorf("Expected EnsureUniqueSlug() to produce result of '%s', got '%s'", "test", actual)
+	}
+
+	// Test simple
+	db.Rows = &database.MockJournal_SingleRow{}
+	db.ExpectedArgument = "test"
+	actual = js.EnsureUniqueSlug("test", 0)
+	if actual != "test-1" {
+		t.Errorf("Expected EnsureUniqueSlug() to produce result of '%s', got '%s'", "test-1", actual)
+	}
+
+	db.Rows = &database.MockJournal_SingleRow{}
+	db.ExpectedArgument = "test-2"
+	actual = js.EnsureUniqueSlug("test", 2)
+	if actual != "test-3" {
+		t.Errorf("Expected EnsureUniqueSlug() to produce result of '%s', got '%s'", "test-3", actual)
+	}
+}
+
 func TestJournals_FetchAll(t *testing.T) {
 
 	// Test error
@@ -123,6 +152,7 @@ func TestJournals_FindBySlug(t *testing.T) {
 
 func TestJournals_Save(t *testing.T) {
 	db := &database.MockSqlite{Result: &database.MockResult{}}
+	db.Rows = &database.MockRowsEmpty{}
 	gs := &database.MockGiphyExtractor{}
 	container := &app.Container{Db: db}
 	js := Journals{Container: container, Gs: gs}

--- a/journal.go
+++ b/journal.go
@@ -17,7 +17,7 @@ import (
 )
 
 func main() {
-	const version = "0.2"
+	const version = "0.2.1"
 
 	// Command line flags
 	var (

--- a/journal_test.go
+++ b/journal_test.go
@@ -126,7 +126,7 @@ func TestApiV1Single_NotFound(t *testing.T) {
 func TestApiV1Create(t *testing.T) {
 	fixtures(t)
 
-	request, err := http.NewRequest("POST", server.URL+"/api/v1/post", strings.NewReader(`{"title":"Test 4","date":"2018-06-01T00:00:00Z","content":"<p>Test 4!</p>"}`))
+	request, err := http.NewRequest("PUT", server.URL+"/api/v1/post", strings.NewReader(`{"title":"Test 4","date":"2018-06-01T00:00:00Z","content":"<p>Test 4!</p>"}`))
 
 	res, err := http.DefaultClient.Do(request)
 
@@ -151,7 +151,7 @@ func TestApiV1Create(t *testing.T) {
 func TestApiV1Create_InvalidRequest(t *testing.T) {
 	fixtures(t)
 
-	request, err := http.NewRequest("POST", server.URL+"/api/v1/post", nil)
+	request, err := http.NewRequest("PUT", server.URL+"/api/v1/post", nil)
 
 	res, err := http.DefaultClient.Do(request)
 
@@ -167,7 +167,7 @@ func TestApiV1Create_InvalidRequest(t *testing.T) {
 func TestApiV1Create_MissingData(t *testing.T) {
 	fixtures(t)
 
-	request, err := http.NewRequest("POST", server.URL+"/api/v1/post", strings.NewReader(`{"title":"Test 4"}`))
+	request, err := http.NewRequest("PUT", server.URL+"/api/v1/post", strings.NewReader(`{"title":"Test 4"}`))
 
 	res, err := http.DefaultClient.Do(request)
 
@@ -183,7 +183,7 @@ func TestApiV1Create_MissingData(t *testing.T) {
 func TestApiV1Update(t *testing.T) {
 	fixtures(t)
 
-	request, err := http.NewRequest("PUT", server.URL+"/api/v1/post/test", strings.NewReader(`{"title":"A different title"}`))
+	request, err := http.NewRequest("POST", server.URL+"/api/v1/post/test", strings.NewReader(`{"title":"A different title"}`))
 
 	res, err := http.DefaultClient.Do(request)
 
@@ -208,7 +208,7 @@ func TestApiV1Update(t *testing.T) {
 func TestApiV1Update_NotFound(t *testing.T) {
 	fixtures(t)
 
-	request, err := http.NewRequest("PUT", server.URL+"/api/v1/post/random", strings.NewReader(`{"title":"A different title"}`))
+	request, err := http.NewRequest("POST", server.URL+"/api/v1/post/random", strings.NewReader(`{"title":"A different title"}`))
 
 	res, err := http.DefaultClient.Do(request)
 
@@ -224,7 +224,7 @@ func TestApiV1Update_NotFound(t *testing.T) {
 func TestApiV1Update_InvalidRequest(t *testing.T) {
 	fixtures(t)
 
-	request, err := http.NewRequest("PUT", server.URL+"/api/v1/post/test", nil)
+	request, err := http.NewRequest("POST", server.URL+"/api/v1/post/test", nil)
 
 	res, err := http.DefaultClient.Do(request)
 

